SST Partitioner interface that allows to split SST files

koldat opened the pull request:
SST Partitioner interface that allows to split SST files during compactions. 

It basically instruct compaction to create a new file when needed. When one is using well defined prefixes and prefixed way of defining tables it is good to define also partitioning so that promotion of some SST file does not cover huge key space on next level (worst case complete space).

yiwu-arbug commented on db/compaction/compaction.h:
```diff
@@ -255,6 +256,9 @@ class Compaction {
   std::unique_ptr<CompactionFilter> CreateCompactionFilter() const;
 
+  // Create a SstPartitioner from sst_partitioner_factory
+  std::unique_ptr<SstPartitioner> CreateSstPartitioner() const;
```
also update options/cf_options.(h|cc) accordingly.

yiwu-arbug commented on db/compaction/compaction.cc:
```diff
@@ -340,6 +344,13 @@ bool Compaction::IsTrivialMove() const {
+
+    if (partitioner.get() != nullptr) {
+      partitioner->Reset(file->smallest.user_key());
+      if (partitioner->ShouldPartition(file->largest.user_key())) {
```
what's the use case for splitting SST when there's opportunity for trivial move? If you thing there's a valid case, can we add an option for it?

koldat replied:
Oh, you are right there can still be case where there is no need to split for some files. I will add some "CanDoTrivialMove(key_form, key_to)" function

koldat replied:
done

yiwu-arbug commented on db/compaction/compaction_job.cc:
```diff
@@ -971,16 +980,19 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
-      output_file_ended = true;
+    if (!output_file_ended && c_iter->Valid()) {
+      if (((partitioner.get() != nullptr &&
+            partitioner->ShouldPartition(c_iter->key())) ||
```
Is the key passing to `partitioner` an user key or internal key? I think it is better to let the `partitioner` only deal with user key.

koldat replied:
It is user key. Internal key would be c_iter->ikey();

yiwu-arbug replied:
There's also `c_iter->user_key()`. I think both `c_iter->key()` and `c_iter->ikey()` are internal key, except the `ikey` is in parsed form.

koldat replied:
done

yiwu-arbug commented on db/compaction/compaction_job.cc:
```diff
@@ -1718,4 +1730,56 @@ void CompactionJob::LogCompaction() {
+    if (last_key_.empty()) {
+      return false;
+    }
+    Slice key_fixed(key.data_, std::min(key.size_, len_));
```
why `last_key_` is never get updated?

koldat replied:
last_key is updated in Reset

yiwu-arbug commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,74 @@
+  virtual std::unique_ptr<SstPartitioner> CreatePartitioner(
+      const SstPartitioner::Context& context) const = 0;
+
+  virtual SstPartitionerFactory* Clone() const = 0;
```
I think the method should be implemented in user code and no need to be an rocksdb interface.

koldat replied:
That clone is needed for Java. It needs to make a copy.

yiwu-arbug replied:
I know nothing about java. But maybe you could make copy out of a shared_ptr instead.

koldat replied:
done

yiwu-arbug commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,74 @@
+  // Called with key that is right after the key that was stored into the SST
+  // Returns true of partition boundary was detected and compaction should
+  // create new file.
+  virtual bool ShouldPartition(const Slice& key) = 0;
```
shall this method also take current output file size as parameter, so the partitioner can also split SST by size?

koldat replied:
Split by size is functionality that compaction knows and there is a lot of options around it. Do we really want to introduce another way? Isn't it confusing? Main purpose of this change is to split based on data.

yiwu-arbug replied:
For our use case splitting SST is just a heuristic, not a mandatory. I want to be more flexible in that I can do "split SST based on key if the file is not too small or too large, otherwise override the behavior by size".

yiwu-arbug replied:
It mean to say "please partition now".

koldat replied:
It is "please partition now". I have put there two parameters to make it easier (removed reset)

yiwu-arbug commented on db/compaction/compaction.h:
```diff
@@ -255,6 +256,9 @@ class Compaction {
   std::unique_ptr<CompactionFilter> CreateCompactionFilter() const;
 
+  // Create a SstPartitioner from sst_partitioner_factory
+  std::unique_ptr<SstPartitioner> CreateSstPartitioner() const;
```
update `ColumnFamilyOptions::Dump` accordingly?

koldat replied:
done

yiwu-arbug commented on db/compaction/compaction_job.cc:
```diff
@@ -945,6 +950,10 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
     sub_compact->num_output_records++;
 
+    if (partitioner.get() != nullptr) {
+      partitioner->Reset(key);
```
Can we merge the `Reset` and `ShouldPartition` call into one method call? (echoing the same comment from https://github.com/facebook/rocksdb/pull/5201#discussion_r278659718)

koldat replied:
It is possible, but then the loop will need to copy previous key. Right now the partitioner can remember only the bit it needs. It is hard to say if that copy will take more time, but I think virtual call is pretty cheap compared to memory copy of X number of bytes. So should I do that?

yiwu-arbug replied:
good point about memcopy. But it looks like in the current approach the `Reset` needs a memcopy as well. The `key` is a slice obtain from `c_iter`, which lifetime is ended after `c_iter->Next()` is called. The partitioner needs to make memcopy to keep it. I'm not sure why ASAN test don't complain. 

In case I didn't misread the code and memcopy is inevitable, I think we can change to the following flow:
```
for KV in c_iter {
    if (partitioner->ShouldPartition(next_key)) {
         output_file_ended = true;
    }
    if (output_file_ended) {
         partitioner->Reset(next_key);
    }
}
``` 
That way the partitioner is invoke (#keys + #files) times, and it can decide on its own whether to memcopy keys per key or per file.

koldat replied:
reset removed and should partition is with two arguments

koldat commented on db/compaction/compaction_job.cc:
```diff
@@ -1718,4 +1730,56 @@ void CompactionJob::LogCompaction() {
   }
 }
 
+class SstPartitionerFixedPrefix : public SstPartitioner {
```
Sure

koldat replied:
done

koldat commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,74 @@
+  virtual const char* Name() const = 0;
+};
+
+extern SstPartitionerFactory* NewSstPartitionerFixedPrefixFactory(
```
I am not sure which one is correct (I have copied approach from other factory). There are some functions returning shared_ptr and some raw pointers.  What is the approach in rocksdb currently? Everything shared_ptr?

yiwu-arbug replied:
I think the newer code use shared_ptr more and older code use raw pointer more. rocksdb team folks can probably give a more precise answer. My personal choice is to use smart pointer as much as possible whenever ownership is involved.

koldat replied:
done

koldat commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,74 @@
+    bool is_manual_compaction;
+    // Output level for this compaction
+    int output_level;
+  };
```
If it is available on all places it is used then yes. I will take a look.

yiwu-arbug replied:
I think passing `compaction.smallest_user_key` and `compaction.largest_user_key` should do.

koldat replied:
done

koldat commented:
@yiwu-arbug thanks for looking on the PR.

mrambacher commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,74 @@
+#include <memory>
+#include <string>
+
+#include "rocksdb/rocksdb_namespace.h"
```
Should this code be limited by ROCKSDB_LITE or not?  I do not know what the plans for LITE is or not and if this is necessary in LITE mode or should be skipped.

koldat replied:
I do not know. Functionality is like 20 lines of code. Please let me know if I should not include in LITE

mrambacher commented on java/rocksjni/options.cc:
```diff
@@ -3621,6 +3635,19 @@ void Java_org_rocksdb_ColumnFamilyOptions_setTableFactory(
+    JNIEnv*, jobject, jlong jhandle, jlong factory_handle) {
+  auto* options =
+      reinterpret_cast<ROCKSDB_NAMESPACE::ColumnFamilyOptions*>(jhandle);
+  auto* factory = reinterpret_cast<ROCKSDB_NAMESPACE::SstPartitionerFactory*>(
```
I think this should be a std::shared_ptr<...> for consistency with how the other Java code behaves.  And this will eliminate the need for the Clone method as well.

koldat replied:
It is now as shared_ptr only.

mrambacher commented on db/compaction/compaction_job.cc:
```diff
@@ -1718,4 +1730,56 @@ void CompactionJob::LogCompaction() {
+      return false;
+    }
+    Slice key_fixed(key.data_, std::min(key.size_, len_));
+    return key_fixed.compare(last_key_) != 0;
```
Shouldn't this be based on a Comparator rather than raw bytes?

koldat replied:
We compare prefix (number of bytes) for equality. I do not see much benefit to use comparator for this partitioner.

koldat commented on java/rocksjni/sst_partitioner.cc:
```diff
@@ -0,0 +1,55 @@
+    JNIEnv*, jobject /*jobj*/, jlong jhandle) {
+  auto* ptr = reinterpret_cast<
+      std::shared_ptr<ROCKSDB_NAMESPACE::SstPartitionerFactory>*>(jhandle);
+  auto* copy = ptr->get()->Clone();
```
clone removed. I used only shared_ptr now.

koldat commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,74 @@
+  virtual bool ShouldPartition(const Slice& key) = 0;
+
+  // Called for key that was stored into the SST
+  virtual void Reset(const Slice& key) = 0;
```
reset removed.

adamretter commented:
@koldat @mrambacher when you guys are happy with the C++ API and functionality, let me know and I will review the Java stuff :-)

koldat commented:
I have rebased to master head and squashed all changesets to make it clean. I have also added current file size to ShouldPartition that I have missed.

yiwu-arbug commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,81 @@
+  // Context information of a compaction run
+  struct Context {
+    // Does this compaction run include all data files
+    bool is_full_compaction_;
```
nit: public members do not end with underscore

koldat replied:
fixed

yiwu-arbug commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,72 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
For new files not sure if we need the leveldb header.

koldat replied:
removed (as well as in header)

yiwu-arbug commented on include/rocksdb/sst_partitioner.h:
```diff
@@ -0,0 +1,81 @@
+
+  // Returns true if partition boundary was detected and compaction should
+  // create new file.
+  virtual bool ShouldPartition(const Slice& last_user_key,
```
Shall we create a new options struct for `ShouldPartition` parameter, so that the interface is extendable?

koldat replied:
I am not sure I understand. We have the Context there already. If you want to pass struct then we should place there also prev and current key there. Is that what you mean? I have also fixed documentation to make the purpose easier to understand.

yiwu-arbug replied:
Yes, that's what I mean. It makes the API easier to extend later.

koldat replied:
I have changed to   
`virtual PartitionerResult ShouldPartition(const PartitionerRequest& request) = 0;`

yiwu-arbug commented on db/compaction/compaction_job.cc:
```diff
@@ -914,6 +916,11 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
 
+  std::unique_ptr<SstPartitioner> partitioner =
+      sub_compact->compaction->CreateSstPartitioner();
+  SstPartitioner* partitioner_ptr = partitioner.get();
```
why adding a new variable instead of calling `partitioner->XXX()` directly?

koldat replied:
Right, fixed.

yiwu-arbug commented on db/compaction/compaction_job.cc:
```diff
@@ -972,20 +979,29 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
         reinterpret_cast<void*>(
             const_cast<std::atomic<bool>*>(manual_compaction_paused_)));
+    if (partitioner_ptr != nullptr) {
+      last_key_for_partitioner.assign(c_iter->user_key().data_,
```
I still want to keep a `Reset()` method (or name it `OnOutputFileEnded()`) to signal new output file created. Also that way we don't need to pass last key to `ShouldPartition()`, thus saving a memcopy. What do you think?

koldat replied:
I was thinking about that and memcpy will not be noticeable compared to everything else. Keys are small. Most partitioners will have to keep last anyways so the copy will happen. Fixed partitioner can potentially copy less, but this makes everything much easier to read and maintain I think. We are doing this check for every pair of keys in compaction so that Reset was called same amount of times as ShouldPartition. Memcpy can be at least inlined.

When ShouldPartition return true, compaction will always split the file so there is no need to create event OnOutputFileEnded I think.

yiwu-arbug replied:
I want to keep the `Reset()` for the following reasons:
1. currently SST split can cause by the partitioner and other conditions (e.g. `max_output_file_size`). I want to leave a mechanism notify the partitioner when SST split actually happen.
2. potentially it allow us to change API to make `ShouldPartition` just a hint.
3. reduce memcpy

yiwu-arbug replied:
After a second thought I think the `Reset` method is not needed now and can be added later.

yiwu-arbug commented on options/options.cc:
```diff
@@ -122,6 +123,9 @@ void ColumnFamilyOptions::Dump(Logger* log) const {
       log, "       Options.compaction_filter_factory: %s",
       compaction_filter_factory ? compaction_filter_factory->Name() : "None");
+  ROCKS_LOG_HEADER(
+      log, "       Options.sst_partitioner_factory: %s",
```
nit: the log line should aligned at the semicolon.

koldat replied:
Fixed. I have kept the other that were not aligned like compaction_filter_factory to not cause some regression.

yiwu-arbug commented on db/compaction/compaction_job.cc:
```diff
@@ -972,20 +978,29 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
+            partitioner->ShouldPartition(
+                last_key_for_partitioner, c_iter->user_key(),
+                sub_compact->current_output_file_size)) ||
+           (sub_compact->compaction->output_level() != 0 &&
```
SST partitioner should not be called when `output_level = 0`, as the intent of intra-L0 compaction is to combine L0 files. It also confuse the LSM consistency check when `force_consistency_checks=true`.

koldat replied:
Good point. fixed

koldat commented:
I have rebased and put latest review changes. I have changed the signature to:
`  virtual PartitionerResult ShouldPartition(
      const PartitionerRequest& request) = 0;`
Result is enum where we can add later kPartitionMaybe, for now there is no/must. Request has same arguments as before, but is extendable (additional fields can be added without API change).

yiwu-arbug commented:
@siying

yiwu-arbug commented:
@siying @ajkr

siying commented on include/rocksdb/options.h:
```diff
@@ -308,6 +310,14 @@ struct ColumnFamilyOptions : public AdvancedColumnFamilyOptions {
+  // If non-nullptr, use the specified factory for a function to determine the
+  // partitioning of sst files. This helps compaction to split the files
+  // on interesting boundaries (key prefixes) to make propagation of sst
+  // files less write amplifying (covering the whole key space).
```
Can we comment and say it's a experimental feature for now?

siying commented:
@koldat if you rebase against master and add a comment for the feature is experimental, we can try to merge it.

siying commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,70 @@
+ public:
+  SstPartitionerFixedPrefix(size_t len) : len_(len) {}
+
+  virtual ~SstPartitionerFixedPrefix(){};
```
```
db/compaction/sst_partitioner.cc:17:41: warning: extra ';' after member function definition
  virtual ~SstPartitionerFixedPrefix(){};
                                        ^
```
and
```
db/compaction/sst_partitioner.cc:17:11: warning: '~SstPartitionerFixedPrefix' overrides a destructor but is not marked 'override'
  virtual ~SstPartitionerFixedPrefix(){};
          ^

include/rocksdb/sst_partitioner.h:46:11: note: overridden virtual function is here
  virtual ~SstPartitioner(){};
          ^
```

koldat replied:
Fixed

siying commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,70 @@
+  size_t len_;
+};
+
+class SstPartitionerFixedPrefixFactory : public SstPartitionerFactory {
```
"Struct and class definitions should either be defined in a header file, declared in a header file with the same base name and path as the definition file, or in an anonymous namespace to prevent possible shadowing of other classes."

koldat replied:
Fixed

siying commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,70 @@
+
+class SstPartitionerFixedPrefix : public SstPartitioner {
+ public:
+  SstPartitionerFixedPrefix(size_t len) : len_(len) {}
```
The linter asks us to add "explicit".

koldat replied:
Fixed

siying commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,70 @@
+
+std::shared_ptr<SstPartitionerFactory> NewSstPartitionerFixedPrefixFactory(
+    size_t prefix_len) {
+  return std::shared_ptr<SstPartitionerFactory>(
```
The linter asks us to use make_shared

koldat replied:
Fixed

siying commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,70 @@
+
+class SstPartitionerFixedPrefixFactory : public SstPartitionerFactory {
+ public:
+  SstPartitionerFixedPrefixFactory(size_t len) : len_(len) {}
```
Same here. add explicit.

koldat replied:
Fixed

siying commented on db/compaction/sst_partitioner.cc:
```diff
@@ -0,0 +1,70 @@
+
+namespace ROCKSDB_NAMESPACE {
+
+class SstPartitionerFixedPrefix : public SstPartitioner {
```
"Struct and class definitions should either be defined in a header file, declared in a header file with the same base name and path as the definition file, or in an anonymous namespace to prevent possible shadowing of other classes."

koldat replied:
Fixed

yiwu-arbug commented:
@koldat ping

koldat commented:
@yiwu-arbug I'll do next week. Was on vacation, sorry.

koldat commented:
@siying I have done all review changes, squashed and rebased to master.

The only question I have is about crash test. You want me to write new test in db_stress using sst_partitioner? It would be quite big task, because we need to define first what should be outcome of such a test. The partitioning is always application specific. On the other hand the only impact on compaction is just "two lines" and in case it is not enabled there is no impact. Please let me know if I should go and create something.

siying commented:
@koldat there can be multiple levels on the partitioned about what we can validate. From a minimal level, we can just apply a prefix partitioned, enable it in stress test by say 1/3 of the time, and watch it doesn't generate wrong query results or crash the service. That's already very helpful. Validating logic correctness is lower priority. Again, I still hope the stress test to be done, but can be done as a follow-up PR.

siying commented:
@koldat sorry but after rebasing the branch to latest master, it seems to show more errors. Can you help fix them?

koldat commented:
@siying I have removed your merge, rebased to master and fixed the build.

siying commented:
Thank you for your contribution!
